src/haproxy.py

```diff
@@ -79,19 +83,27 @@ def is_active(self) -> bool:
         """
         return systemd.service_running(APT_PACKAGE_NAME)
 
-    def _render_haproxy_config(self, config: CharmConfig) -> None:
+    def _render_haproxy_config(self, config: CharmConfig, http_provider: HTTPProvider) -> None:
         """Render the haproxy configuration file.
 
         Args:
-            config: charm config
+            config: The charm config.
+            http_provider: The http interface provider.
         """
         with open("templates/haproxy.cfg.j2", "r", encoding="utf-8") as file:
-            template = Template(file.read(), keep_trailing_newline=True)
-        rendered = template.render(config_global_max_connection=config.global_max_connection)
+            template = Template(
+                file.read(), keep_trailing_newline=True, trim_blocks=True, lstrip_blocks=True
+            )
```

Comment:
Would https://docs.python.org/3/library/pathlib.html#pathlib.Path.read_text work here? Pathlib is a more modern file system interface in Python

Reply:
In a PR further down the line I switch to using loaders to be able to manage inheritance, so since this code will be replaced soon, I'll resolved this comment if that's ok.